plugin/evm/state/state.go

```diff
@@ -0,0 +1,370 @@
+		if err != nil {
+			return err
+		}
+		internalBlock, err := getBlock(lastAcceptedID)
+		if err != nil {
+			return err
+		}
+		lastAcceptedBlock = &BlockWrapper{
+			Block:  internalBlock,
+			status: choices.Accepted,
+		}
+	} else {
+		genesisID := genesisBlock.ID()
+		if err := c.internalDB.Put(LastAcceptedKey, genesisID[:]); err != nil {
+			return err
+		}
+		if err := c.acceptedBlocksByHeight.Put(heightKey(genesisBlock.Height()), genesisID[:]); err != nil {
+			return err
+		}
+		lastAcceptedBlock = &BlockWrapper{
+			Block:  genesisBlock,
+			status: choices.Accepted,
+		}
+	}
+	lastAcceptedBlock.state = c
+	c.lastAcceptedBlock = lastAcceptedBlock
+	// Put the last accepted block in the decided block cache to start.
+	c.decidedBlocks.Put(lastAcceptedBlock.ID(), lastAcceptedBlock)
+
+	return nil
+}
+
+// FlushCaches flushes each block cache completely.
+func (c *ChainState) FlushCaches() {
+	c.decidedBlocks.Flush()
+	c.missingBlocks.Flush()
+	c.unverifiedBlocks.Flush()
+}
+
+// ExternalDB returns a database to be used external to ChainState
+// The returned database must handle batching and atomicity by itself except for
+// any operations that take place during block decisions. Any database operations
+// that occur during block decisions (Accept/Reject) will be automatically batched.
+func (c *ChainState) ExternalDB() database.Database { return c.externalDB }
+
+// GetBlock returns the BlockWrapper as snowman.Block corresponding to [blkID]
+func (c *ChainState) GetBlock(blkID ids.ID) (snowman.Block, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if blk, ok := c.getCachedBlock(blkID); ok {
+		return blk, nil
+	}
+
+	if _, ok := c.missingBlocks.Get(blkID); ok {
+		return nil, errUnknownBlock
+	}
+
+	c.lock.Unlock()
+	blk, err := c.getBlock(blkID)
+	c.lock.Lock()
+	if err != nil {
+		c.missingBlocks.Put(blkID, struct{}{})
+		return nil, err
+	}
+
+	// Since this block is not in processing, we can add it as
+	// a non-processing block to the correct cache.
+	return c.addNonProcessingBlock(blk)
+}
+
+// getCachedBlock checks the caches for [blkID] by priority. Returning
+// true if [blkID] is found in one of the caches.
+func (c *ChainState) getCachedBlock(blkID ids.ID) (snowman.Block, bool) {
+	if blk, ok := c.processingBlocks[blkID]; ok {
+		return blk, true
+	}
+
+	if blk, ok := c.decidedBlocks.Get(blkID); ok {
+		return blk.(snowman.Block), true
+	}
+
+	if blk, ok := c.unverifiedBlocks.Get(blkID); ok {
+		return blk.(snowman.Block), true
+	}
+
+	return nil, false
+}
+
+// GetBlockInternal returns the internal representation of [blkID]
+func (c *ChainState) GetBlockInternal(blkID ids.ID) (Block, error) {
+	wrappedBlk, err := c.GetBlock(blkID)
+	if err != nil {
+		return nil, err
+	}
+
+	return wrappedBlk.(*BlockWrapper).Block, nil
+}
+
+// getBlockIDAtHeight returns the blkID accepted at [height] if there is
+// one.
+// assumes the lock is held.
+func (c *ChainState) getBlockIDAtHeight(height uint64) (ids.ID, error) {
```

Comment:
Why does this exist? Shouldn't this just get Geth's canonical hash?

Reply:
Going to remove in favor of geth's canonical hash.